Sources/Plasma/NucleusLib/pnAsyncCoreExe/pnAceSocket.cpp

```diff
@@ -0,0 +1,561 @@
+            threadCount = kMaxWorkerThreads;
+            LogMsg(kLogError, "kMaxWorkerThreads too small!");
+        }
+
+        // create IO worker threads
+        for (unsigned int thread = 0; thread < threadCount; thread++) {
+            fThreadHandles[thread] = std::thread([this] {
+#ifdef USE_VLD
+                VLDEnable();
+#endif
+                PerfAddCounter(kAsyncPerfThreadsTotal, 1);
+                PerfAddCounter(kAsyncPerfThreadsCurr, 1);
+
+                // This can be run concurrently from several threads
+                fContext.run();
+
+                PerfSubCounter(kAsyncPerfThreadsCurr, 1);
+            });
+        }
+    }
+
+    void Destroy(unsigned exitThreadWaitMs)
+    {
+        fWorkGuard.reset();
+
+        for (std::thread& thread : fThreadHandles) {
+            if (thread.joinable())
+                AsyncThreadTimedJoin(thread, exitThreadWaitMs);
+        }
+
+        // Ensure the event loop exits without processing any more tasks,
+        // in case the threads take more than exitThreadWaitMs to finish
+        fContext.stop();
+    }
+};
+
+struct ConnectOperation
+{
+    tcp::socket             fSock;
+    plNetAddress            fRemoteAddr;
+    AsyncCancelId           fCancelId;
+    FAsyncNotifySocketProc  fNotifyProc;
+    void*                   fParam;
+    std::vector<uint8_t>    fConnectBuffer;
+    unsigned int            fConnectionType;
+
+    ConnectOperation(asio::io_context& context)
+        : fSock(context), fCancelId(), fParam(), fConnectionType()
+    { }
+};
+
+struct WriteOperation
+{
+    unsigned int            fAllocSize;
+    AsyncNotifySocketWrite  fNotify;
+    unsigned                queueTimeMs;
+
+    WriteOperation() : fAllocSize() { }
+
+    asio::const_buffer AsBuffer() const
+    {
+        return asio::buffer(fNotify.buffer + fNotify.bytesCommitted, fNotify.bytes - fNotify.bytesCommitted);
+    }
+};
+
+struct AsyncSocketStruct
+{
+    std::recursive_mutex        fCritsect;
+    tcp::socket                 fSock;
+    FAsyncNotifySocketProc      fNotifyProc;
+    void*                       fParam;
+    unsigned int                fConnectionType;
+    void*                       fUserState;
+    uint8_t                     fBuffer[kAsyncSocketBufferSize];
+    size_t                      fBytesLeft;
+    std::list<WriteOperation *> fWriteOps;
+    unsigned                    initTimeMs;
+    unsigned                    closeTimeMs;
+
+    AsyncSocketStruct(ConnectOperation* op)
+        : fSock(std::move(op->fSock)), fNotifyProc(op->fNotifyProc),
+          fParam(op->fParam), fConnectionType(op->fConnectionType),
+          fUserState(), fBuffer(), fBytesLeft(), initTimeMs(), closeTimeMs()
+    { }
+};
+
+static AsyncIoPool*                 s_ioPool;
+
+static std::recursive_mutex         s_connectCrit;
+static std::list<ConnectOperation*> s_connectList;
+static unsigned                     s_nextConnectCancelId = 1;

```

Comment:
This should be `uintptr_t` rather than `unsigned`. I'm seeing warnings about casting it to a `AsyncCancelId` (which is effectively a `void*`) because of size differences.

Reply:
Fixed!